src/arch/x86_64.rs

```diff
@@ -55,6 +66,37 @@ impl BootInfo for RustbootBootInfo {
         vga::writer()
     }
 
+    fn framebuffer(&self) -> Option<Self::Framebuffer> {
+        let framebuffer = self.inner.framebuffer.as_ref()?;
+
+        static FRAMEBUFFER: InitOnce<spin::Mutex<Framebuffer>> = InitOnce::uninitialized();
+        Some(
+            FRAMEBUFFER
+                .get_or_else(move || {
+                    let info = framebuffer.info();
+                    let buf = framebuffer.buffer();
```

Comment:
Given that we're mutating through this reference, shouldn't we use `buffer_mut`?

I suppose we don't own the actual `BootInfo` struct so we can't get a mutable reference to it here? Feels sketchy, especially because we're given a `&'static mut` reference to the `BootInfo` by the bootloader, but I suppose changing it would be tricky, and looking at the `BootInfo` struct, it seems like the `FrameBuffer` value is the only one which requires mutable access for meaningful operations.

Perhaps we should transfer the `FrameBuffer` out of the `BootInfo` while we have mutable access to it in `arch_entry` and pass ownership of the struct into our `Framebuffer` type, so we can use it directly (or I suppose pass a mutable reference into `kernel_main`)? I don't love doing more stuff before we've set up things like interrupt handlers though.

Reply:
@mystor is https://github.com/hawkw/mycelium/pull/113/commits/270c2351d304a44e2d6f734b179d3d533a7dd35f any less sketchy? it now involves a gross `transmute`, so i'm not sure if that's any better...

Reply:
hmm, but without the framebuffer we can't log anything or display an oops screen on fautls? so i think we should just "try really hard not to fault here"

Reply:
okay, https://github.com/hawkw/mycelium/pull/113/commits/82c1d9aafd68ee1ed9247a066fab3e744246896a should be _actually_ not sketchy. it required some weird refactoring and makes the interfaces somewhat stranger, but it actually does everything we want, and we _always_ get the `&'static mut [u8]` from the bootloader's framebuffer struct, and we don't have a weird transmute...